<a href="https://colab.research.google.com/github/Heechan-Kim-2869/OSS_11/blob/main/%EC%9E%90%EC%82%B4%EB%A5%A0_%EA%B5%AC%ED%95%98%EA%B8%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
파일 분석해줘

Here is all the data you need:
"Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv"

## Data loading

### Subtask:
Load the provided CSV file into a Pandas DataFrame.


**Reasoning**:
Load the provided CSV file into a pandas DataFrame.



In [1]:
import pandas as pd

df = pd.read_csv('Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv')
display(df.head())

INDICATOR                                               UNIT  \
0  Death rates for suicide  Deaths per 100,000 resident population, age-ad...   
1  Death rates for suicide  Deaths per 100,000 resident population, age-ad...   
2  Death rates for suicide  Deaths per 100,000 resident population, age-ad...   
3  Death rates for suicide  Deaths per 100,000 resident population, age-ad...   
4  Death rates for suicide  Deaths per 100,000 resident population, age-ad...   

   UNIT_NUM STUB_NAME  STUB_NAME_NUM   STUB_LABEL  STUB_LABEL_NUM  YEAR  \
0         1     Total              0  All persons             0.0  1950   
1         1     Total              0  All persons             0.0  1960   
2         1     Total              0  All persons             0.0  1970   
3         1     Total              0  All persons             0.0  1980   
4         1     Total              0  All persons             0.0  1981   

   YEAR_NUM       AGE  AGE_NUM  ESTIMATE FLAG  
0         1  All ages      0.0      13.2  NaN  
1         2  All ages      0.0      12.5  NaN  
2         3  All ages      0.0      13.1  NaN  
3         4  All ages      0.0      12.2  NaN  
4         5  All ages      0.0      12.3  NaN

## Data exploration

### Subtask:
Explore the loaded data to understand its structure and key characteristics.


**Reasoning**:
Examine the DataFrame's shape, data types, and unique values in specified columns to understand its structure and key characteristics.



In [2]:
# Determine the shape of the DataFrame.
print(f"Shape of the DataFrame: {df.shape}")

# Identify the data type of each column.
print("\nData types of each column:")
print(df.dtypes)

# Identify unique values and their counts for specified columns.
for col in ['Sex', 'Race', 'Hispanic Origin', 'Age Group']:
    if col in df.columns:
        print(f"\nUnique values and counts for '{col}':")
        try:
            print(df[col].value_counts())
        except TypeError:
            print(f"Column '{col}' has a non-categorical data type and cannot be used with value_counts().")
    else:
        print(f"\nColumn '{col}' not found in the DataFrame.")

Shape of the DataFrame: (6390, 13)

Data types of each column:
INDICATOR          object
UNIT               object
UNIT_NUM            int64
STUB_NAME          object
STUB_NAME_NUM       int64
STUB_LABEL         object
STUB_LABEL_NUM    float64
YEAR                int64
YEAR_NUM            int64
AGE                object
AGE_NUM           float64
ESTIMATE          float64
FLAG               object
dtype: object

Column 'Sex' not found in the DataFrame.

Column 'Race' not found in the DataFrame.

Column 'Hispanic Origin' not found in the DataFrame.

Column 'Age Group' not found in the DataFrame.


## Data analysis

### Subtask:
Calculate descriptive statistics for the 'Death Rate' column, grouped by relevant columns.


**Reasoning**:
Calculate descriptive statistics for the 'ESTIMATE' column, grouped by 'YEAR' and 'AGE'.



In [3]:
import pandas as pd

# Calculate overall descriptive statistics for 'ESTIMATE'.
overall_stats = df['ESTIMATE'].describe()
print("Overall Descriptive Statistics for 'ESTIMATE':\n", overall_stats)

# Group data by 'YEAR' and calculate descriptive statistics.
yearly_stats = df.groupby('YEAR')['ESTIMATE'].describe()
print("\nDescriptive Statistics for 'ESTIMATE' by Year:\n", yearly_stats)

# Group data by 'AGE' and calculate descriptive statistics.
if 'AGE' in df.columns:
    age_stats = df.groupby('AGE')['ESTIMATE'].describe()
    print("\nDescriptive Statistics for 'ESTIMATE' by Age:\n", age_stats)
else:
    print("\nColumn 'AGE' not found in the DataFrame.")

Overall Descriptive Statistics for 'ESTIMATE':
 count    5484.000000
mean       13.709810
std        11.531805
min         0.300000
25%         5.000000
50%        10.500000
75%        19.500000
max        74.800000
Name: ESTIMATE, dtype: float64

Descriptive Statistics for 'ESTIMATE' by Year:
       count       mean        std  min    25%    50%     75%   max
YEAR                                                              
1950   74.0  17.017568  16.965310  0.3  5.700   9.90  22.050  61.9
1960   73.0  16.171233  14.883431  0.5  5.900  10.20  21.100  61.3
1970   76.0  15.542105  11.538125  0.3  7.775  12.30  20.800  45.8
1980   93.0  14.989247  11.338655  0.3  6.100  11.80  19.900  52.8
1981   81.0  14.904938  11.163645  0.5  6.300  12.20  20.000  54.0
1982   81.0  15.116049  11.817799  0.4  6.200  12.20  20.500  54.5
1983   80.0  15.571250  12.410849  0.6  6.450  13.15  20.700  57.9
1984   81.0  15.640741  12.437833  0.6  6.200  12.70  21.300  56.3
1985  118.0  15.042373  12.539199 

## Summary:

### 1. Q&A

No questions were explicitly asked in the provided data analysis task.  However, the analysis implicitly addresses questions like:

* What is the overall distribution of suicide death rates?
* How do suicide death rates vary by year?
* How do suicide death rates vary by age group?


### 2. Data Analysis Key Findings

* **Dataframe Shape:** The DataFrame has 6390 rows and 13 columns.
* **Missing Columns:** The columns 'Sex', 'Race', 'Hispanic Origin', and 'Age Group' (as specified in the original request), were not found in the DataFrame.
* **Overall Death Rate:** The average suicide death rate ('ESTIMATE') is approximately 13.71, with a standard deviation of 11.53, a minimum of 0.3 and a maximum of 74.8.
* **Death Rate Variation by Year:** Suicide death rates fluctuate across different years, suggesting a time-dependent trend.
* **Death Rate Variation by Age:**  Suicide death rates vary significantly across different age groups, with lower rates in younger age groups (e.g., 10-14) and generally increasing rates with age.


### 3. Insights or Next Steps

* **Investigate Temporal Trends:** Further analyze the yearly trends in suicide death rates to identify potential contributing factors and periods of significant change.  This could involve time series analysis or correlation with external factors.
* **Explore Other Variables:**  Since the originally requested columns ('Sex', 'Race', 'Hispanic Origin', and 'Age Group') are not present, locate these variables in the dataset or seek an alternative dataset that includes them to examine how these demographic factors relate to suicide death rates.


In [12]:
# prompt: /content/Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv
# 연도를 입력하면 해당 년도의 자살률을 나이대별로 출력하는 코드를 짤거야. 1950~2018 사이의 연도를 입력하게 해줘. 연도를 입력하면 데이터에서 "ESTIMATE" column을 조사해서 해당 연도의 평균을 구해줘. 평균치는 반올림한 정수로 나타내줘. 또한 해당 년도의 데이터 중, 나이를 나타내는 데이터가 있다면 따로 수집해서 나이대별로 10살씩 다시 "ESTIMATE" column을 평균을 구해서 출력해줘. 이 평균치는 소숫점 한 자릿수의 실수로 출력해줘.

import pandas as pd

def analyze_suicide_rates(year):
    df = pd.read_csv('/content/Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv')

    if not 1950 <= year <= 2018:
        print("Please enter a year between 1950 and 2018.")
        return

    # Calculate the overall average death rate for the given year
    yearly_data = df[df['YEAR'] == year]
    average_death_rate = round(yearly_data['ESTIMATE'].mean())
    print(f"The average suicide death rate in {year} is: {average_death_rate}")

    # Analyze age-specific death rates
    if 'AGE' in df.columns:  #Check if the 'AGE' column exists
        age_groups = yearly_data['AGE'].unique()
        for age_group in age_groups:
            age_specific_data = yearly_data[yearly_data['AGE'] == age_group]
            avg_rate = age_specific_data["ESTIMATE"].mean()
            print(f"Average death rate for age group {age_group}: {avg_rate:.1f}")
    else:
        print("Age-specific data not available in the dataset.")

# Get user input for the year
year_input = int(input("Enter a year between 1950 and 2018: "))
analyze_suicide_rates(year_input)


Enter a year between 1950 and 2018: 1987
The average suicide death rate in 1987 is: 16
Average death rate for age group All ages: 10.9
Average death rate for age group 10-14 years: 1.5
Average death rate for age group 15-24 years: 11.0
Average death rate for age group 15-19 years: 10.1
Average death rate for age group 20-24 years: 14.9
Average death rate for age group 25-44 years: 13.6
Average death rate for age group 25-34 years: 15.5
Average death rate for age group 35-44 years: 15.1
Average death rate for age group 45-64 years: 13.7
Average death rate for age group 45-54 years: 16.2
Average death rate for age group 55-64 years: 17.2
Average death rate for age group 65 years and over: 22.3
Average death rate for age group 65-74 years: 23.9
Average death rate for age group 75-84 years: 34.3
Average death rate for age group 85 years and over: 41.7


# 해당 년도의 평균 자살률과 나이대별 평균 자살률 구하기

## 데이터 선택
데이터는 미국 정부에서 운영하는 오픈 데이터 사이트를 이용함.

**Death rates for suicide, by sex, race, Hispanic origin, and age: United States**

https://data.gov/

데이터 중, 미국에서 일어난 자살을 성별, 인종, 히스패닉 여부, 나이에 따라 정리된 데이터를 발견하였고, 이 데이터를 이용하여 과제를 수행함.

https://catalog.data.gov/dataset/death-rates-for-suicide-by-sex-race-hispanic-origin-and-age-united-states-020c1

## 데이터 분석

Gemini를 이용하여 데이터에서 어떤 분석을 기대할 수 있을지를 의뢰함. 데이터에서 성별과 인종 등의 나이를 제외한 데이터들은 있는 경우도 있고 없는 경우도 있어, 완벽하게 분석할 수 없다는 사실을 알게됨.

## 코드 만들기

다음은 생성형 AI를 이용해 연도와 나이대에 따른 자살률을 얻기 위해 입력한 값임.

*/content/Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv*

*연도를 입력하면 해당 년도의 자살률을 나이대별로 출력하는 코드를 짤거야. 1950~2018 사이의 연도를 입력하게 해줘. 연도를 입력하면 데이터에서 "ESTIMATE" column을 조사해서 해당 연도의 평균을 구해줘. 평균치는 반올림한 정수로 나타내줘. 또한 해당 년도의 데이터 중, 나이를 나타내는 데이터가 있다면 따로 수집해서 나이대별로 10살씩 다시 "ESTIMATE" column을 평균을 구해서 출력해줘. 이 평균치는 소숫점 한 자릿수의 실수로 출력해줘.*

## 코드 실행 후 입력

코드 실행 시 데이터세트의 연도의 범위를 설정하기 위해 이를 생성형 AI에게 주문했으며, 실행 시 *Please enter a year between 1950 and 2018.*이라는 문구가 출력되며 해당하는 연도를 입력하면 됨.

## 코드 결과

코드의 결과는 해당 연도의 자살률과 나이대별의 자살률의 평균값을 구해서 출력하며, 출력 결과는 다음과 같음.


*  예시, '1987'을 입력.

*Enter a year between 1950 and 2018: 1987
The average suicide death rate in 1987 is: 16
Average death rate for age group All ages: 10.9
Average death rate for age group 10-14 years: 1.5
Average death rate for age group 15-24 years: 11.0
Average death rate for age group 15-19 years: 10.1
Average death rate for age group 20-24 years: 14.9
Average death rate for age group 25-44 years: 13.6
Average death rate for age group 25-34 years: 15.5
Average death rate for age group 35-44 years: 15.1
Average death rate for age group 45-64 years: 13.7
Average death rate for age group 45-54 years: 16.2
Average death rate for age group 55-64 years: 17.2
Average death rate for age group 65 years and over: 22.3
Average death rate for age group 65-74 years: 23.9
Average death rate for age group 75-84 years: 34.3
Average death rate for age group 85 years and over: 41.7*

## 결론

코드가 정상적으로 작동하며, 요청한 대로 해당 연도의 자살률은 정수로 나타내며 나이대별 사망률은 소숫점 한자리의 실수로 나타냄.

In [ ]:
# prompt: /content/Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv
# 연도를 입력하면 해당 년도의 자살률을 나이대별로 출력하는 코드를 짤거야. 1950~2018 사이의 연도를 입력하게 해줘. 연도를 입력하면 데이터에서 "ESTIMATE" column을 조사해서 해당 연도의 평균을 구해줘. 평균치는 반올림한 정수로 나타내줘. 또한 해당 년도의 데이터 중, 나이를 나타내는 데이터가 있다면 따로 수집해서 나이대별로 10살씩 다시 "ESTIMATE" column을 평균을 구해서 출력해줘. 이 평균치는 소숫점 한 자릿수의 실수로 출력해줘.

import pandas as pd

def analyze_suicide_rates(year):
    df = pd.read_csv('/content/Death_rates_for_suicide__by_sex__race__Hispanic_origin__and_age__United_States.csv')

    if not 1950 <= year <= 2018:
        print("Please enter a year between 1950 and 2018.")
        return

    # Calculate the overall average death rate for the given year
    yearly_data = df[df['YEAR'] == year]
    average_death_rate = round(yearly_data['ESTIMATE'].mean())
    print(f"The average suicide death rate in {year} is: {average_death_rate}")

    # Analyze age-specific death rates
    if 'AGE' in df.columns:  #Check if the 'AGE' column exists
        age_groups = yearly_data['AGE'].unique()
        for age_group in age_groups:
            age_specific_data = yearly_data[yearly_data['AGE'] == age_group]
            avg_rate = age_specific_data["ESTIMATE"].mean()
            print(f"Average death rate for age group {age_group}: {avg_rate:.1f}")
    else:
        print("Age-specific data not available in the dataset.")

# Get user input for the year
year_input = int(input("Enter a year between 1950 and 2018: "))
analyze_suicide_rates(year_input)
